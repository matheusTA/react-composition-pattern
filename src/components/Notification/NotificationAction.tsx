import { ButtonHTMLAttributes, ElementType } from 'react'
import { twMerge } from 'tailwind-merge'

interface NotificationActionProps
  extends ButtonHTMLAttributes<HTMLButtonElement> {
  icon: ElementType
}

export function NotificationAction({
  icon: Icon,
  className,
  ...restProps
}: NotificationActionProps) {
  return (
    <button
      className={twMerge(
        'flex h-8 w-8 items-center justify-center rounded bg-zinc-400 hover:bg-zinc-500',
        className,
      )}
      {...restProps}
    >
      <Icon className="h-3 w-3 text-zinc-50" />
    </button>
  )
}
