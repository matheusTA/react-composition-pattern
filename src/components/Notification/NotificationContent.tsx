interface NotificationContentProps {
  text: string
}

export function NotificationContent({ text }: NotificationContentProps) {
  return (
    <div className="flex flex-1 flex-col gap-2">
      <p className="text-sm leading-relaxed text-zinc-600">{text}</p>

      <div className="flex items-center gap-1 text-xxs text-zinc-400">
        <span>Convite</span>
        <span>HÃ¡ 3 min</span>
      </div>
    </div>
  )
}
