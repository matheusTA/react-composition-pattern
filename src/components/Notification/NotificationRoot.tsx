import { ReactNode } from 'react'

interface NotificationRootProps {
  children: ReactNode
}

export function NotificationRoot({ children }: NotificationRootProps) {
  return (
    <div className="flex items-start gap-6 bg-zinc-200 px-8 py-4">
      {children}
    </div>
  )
}
