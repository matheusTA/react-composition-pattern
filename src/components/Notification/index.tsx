import { NotificationRoot } from './NotificationRoot'
import { NotificationIcon } from './NotificationIcon'
import { NotificationContent } from './NotificationContent'
import { NotificationActions } from './NotificationActions'
import { NotificationAction } from './NotificationAction'

export const Notification = {
  Root: NotificationRoot,
  Icon: NotificationIcon,
  Content: NotificationContent,
  Actions: NotificationActions,
  Action: NotificationAction,
}
