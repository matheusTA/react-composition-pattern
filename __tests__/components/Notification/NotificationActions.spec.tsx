import { act, render } from '@testing-library/react'
import { NotificationActions } from '@/components/Notification/NotificationActions'

import '@testing-library/jest-dom'

describe('NotificationActions', () => {
  it('renders NotificationActions unchanged', () => {
    const children = (
      <>
        <button data-testid="button-1">cancel</button>
        <button data-testid="button-2">submit</button>
      </>
    )

    const { container } = render(
      <NotificationActions>{children}</NotificationActions>,
    )

    expect(container).toMatchSnapshot()
  })

  it('renders children correctly', () => {
    const children = (
      <>
        <button data-testid="button-1">cancel</button>
        <button data-testid="button-2">submit</button>
      </>
    )

    const { getByText } = render(
      <NotificationActions>{children}</NotificationActions>,
    )

    const buttonCancel = getByText('cancel')
    const buttonSubmit = getByText('submit')

    expect(buttonCancel).toBeInTheDocument()
    expect(buttonSubmit).toBeInTheDocument()
  })

  it('calls onClick prop when button is clicked', () => {
    const onSubmitMock = jest.fn()

    const children = (
      <>
        <button data-testid="button-2" onClick={onSubmitMock}>
          submit
        </button>
      </>
    )

    const { getByText } = render(
      <NotificationActions>{children}</NotificationActions>,
    )

    const buttonSubmit = getByText('submit')

    act(() => {
      buttonSubmit.click()
    })

    expect(onSubmitMock).toBeCalledTimes(1)
  })

  it('applies the correct CSS class to the container div', () => {
    const children = (
      <>
        <button data-testid="button-1">cancel</button>
        <button data-testid="button-2">submit</button>
      </>
    )

    const { container } = render(
      <NotificationActions>{children}</NotificationActions>,
    )

    const notificationActionsDiv = container.firstChild

    expect(notificationActionsDiv).toBeInTheDocument()
    expect(notificationActionsDiv).toHaveClass('flex gap-2 self-center')
  })
})
