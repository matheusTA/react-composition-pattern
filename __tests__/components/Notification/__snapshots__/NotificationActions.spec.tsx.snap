// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationActions renders NotificationActions unchanged 1`] = `
<div>
  <div
    class="flex gap-2 self-center"
  >
    <button
      data-testid="button-1"
    >
      cancel
    </button>
    <button
      data-testid="button-2"
    >
      submit
    </button>
  </div>
</div>
`;
